src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java

```diff
@@ -101,21 +101,26 @@ private void attestationExpirationCheck(long timerId) {
         // This check is to avoid the attest() function takes longer than 60sec and get called again from this method while attesting.
         if (this.isAttesting) {
             LOGGER.warn("In the process of attesting. Skip re-attest.");
-        } else {
-            Instant currentTime = clock.now();
-            Instant tenMinutesBeforeExpire = attestationTokenExpiresAt.minusSeconds(600);
-
-            if (currentTime.isAfter(tenMinutesBeforeExpire)) {
-                LOGGER.info("Attestation token is 10 mins from the expiry timestamp {}. Re-attest...", attestationTokenExpiresAt);
-                try {
-                    this.isAttesting = true;
-                    attest();
-                } catch (AttestationTokenRetrieverException | IOException e) {
-                    notifyResponseStatusWatcher(401);
-                    LOGGER.info("Re-attest failed: ", e);
-                } finally {
-                    this.isAttesting = false;
-                }
+            return;
+        }
+
+        Instant currentTime = clock.now();
+        Instant tenMinutesBeforeExpire = attestationTokenExpiresAt.minusSeconds(600);
+        if (currentTime.isAfter(tenMinutesBeforeExpire)) {
+            LOGGER.info("Attestation token is 10 mins from the expiry timestamp {}. Re-attest...", attestationTokenExpiresAt);
+            if (!attestationProvider.isReady()) {
+                LOGGER.warn("Attestation provider is not ready. Skip re-attest.");
+                return;
+            }
+
+            try {
+                this.isAttesting = true;
```

Comment:
BTW we may want to put this `this.isAttesting` in the very beginning after the `this.isAttesting` check, otherwise there would be multiple entry issue. (Though this should not be called from multiple thread.)

It's still not thread safe.. up to you whether want to use CompareAndSwap.

Reply:
Good catch. This may be called by the periodic timer from multiple threads, let me fix it.

Reply:
Updated.